### PR TITLE
[Integration][PagerDuty] Support custom fields on services and incidents

### PR DESCRIPTION
**Pull request by shariff-6**
### **User description**
# Description

What - Add support for custom fields on PagerDuty services and incidents

Why - Users need to ingest custom field values attached to PagerDuty services and 
incidents into Port
                                                                                  
How - Added a new includeCustomFields boolean selector option (default false) for
both services and incidents. When enabled, custom fields are fetched per-entity
via asyncio.gather with return_exceptions=True for resilience, and attached as
__custom_fields on each entity. Enrichment is applied in both resync and webhook
paths.

## Type of change

Please leave one option from the following and delete the rest:

- [x] New feature (non-breaking change which adds functionality)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Enhancement

___

### **Description**
- Add custom field support for PagerDuty services and incidents

- Implement parallel fetching via asyncio.gather with error resilience

- Apply enrichment in both resync and webhook processing flows

- Add comprehensive test coverage for custom field operations

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["PagerDuty API"] -->|"fetch custom fields"| B["get_entity_custom_fields()"]
  B -->|"parallel gather"| C["enrich_entities_with_custom_fields()"]
  C -->|"attach __custom_fields"| D["Enriched Entities"]
  E["Resync Flow"] -->|"if includeCustomFields"| C
  F["Webhook Flow"] -->|"if includeCustomFields"| C
  D -->|"yield/return"| G["Port Ingestion"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pagerduty.py</strong><dd><code>Add custom field fetching and enrichment methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/clients/pagerduty.py

<ul><li>Add <code>get_entity_custom_fields()</code> method to fetch custom fields for <br>single entities via API<br> <li> Add <code>enrich_entities_with_custom_fields()</code> method for batch parallel <br>enrichment using asyncio.gather<br> <li> Implement error handling with return_exceptions=True for resilience<br> <li> Attach custom fields as <code>__custom_fields</code> property on each entity</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-79febc4a748111076c6ff542e5813561fd6e51a42ac805e969719e0aa9ebac24">+35/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Integrate custom field enrichment in resync flows</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/main.py

<ul><li>Refactor incident resync to apply custom field enrichment when enabled<br> <li> Refactor service resync to apply custom field enrichment when enabled<br> <li> Consolidate enrichment logic to apply both analytics and custom fields <br>sequentially<br> <li> Simplify yield logic by applying all enrichments before yielding</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-825286bba1aa4733bb5bc03e85b7d108fb57286338387034e0b7ca1f42d72e83">+14/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>incidents.py</strong><dd><code>Apply custom field enrichment in incident webhooks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/webhook_processors/incidents.py

<ul><li>Add custom field enrichment call in incident webhook event handler<br> <li> Apply enrichment conditionally based on <code>include_custom_fields</code> selector <br>flag<br> <li> Maintain existing analytics enrichment logic</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-1aafb6280ac71f7975afd048fe3b1e901baea3dcd6ea1669ede0813572f6f986">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>services.py</strong><dd><code>Apply custom field enrichment in service webhooks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/webhook_processors/services.py

<ul><li>Add import for <code>cast</code> type utility and <code>PagerdutyServiceResourceConfig</code><br> <li> Add custom field enrichment call in service webhook event handler<br> <li> Apply enrichment conditionally based on <code>include_custom_fields</code> selector <br>flag<br> <li> Integrate enrichment within retry loop after oncall user update</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-88b82caabaeab46b64510118523611ebc4b9b25480689578f20c578f19497de2">+11/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>integration.py</strong><dd><code>Add includeCustomFields selector options</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/integration.py

<ul><li>Add <code>include_custom_fields</code> boolean selector field to <br>PagerdutyIncidentResourceConfig<br> <li> Add <code>include_custom_fields</code> boolean selector field to <br>PagerdutyServiceResourceConfig<br> <li> Both fields default to false with descriptive documentation<br> <li> Use camelCase alias <code>includeCustomFields</code> for API compatibility</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-d65cdf1ba10e9067c90d47298598f419fd62a97e0f9e6937f8518b34066e4bca">+10/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_client.py</strong><dd><code>Add custom field client method tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/tests/test_client.py

<ul><li>Add test for <code>get_entity_custom_fields()</code> with valid custom fields <br>response<br> <li> Add test for <code>get_entity_custom_fields()</code> with empty custom fields<br> <li> Add test for <code>enrich_entities_with_custom_fields()</code> batch enrichment<br> <li> Add test for partial enrichment with mixed empty and populated custom <br>fields</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-9909f1f9ed695bc8b6cf7334048069b8073f645654a3d8cf3f9474cbc61e0192">+89/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_webhook_processors.py</strong><dd><code>Add webhook processor tests for custom fields</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/tests/test_webhook_processors.py

<ul><li>Expand incident_resource_config fixture to test all combinations of <br>analytics and custom fields<br> <li> Add new service_resource_config fixture with custom fields parameter <br>variations<br> <li> Add test for service webhook with custom fields enrichment <br>verification<br> <li> Add assertions to verify custom field enrichment is called only when <br>enabled</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2773/files#diff-981889a47c42189a1e3f7b4acbe4bb164cfcb1a51c017b199980253e7804b2d9">+100/-6</a>&nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Review by mk-armah — APPROVED** [review]
Great

File: integrations/pagerduty/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
+## 0.4.48 (2026-02-12)
```
Comment (matan84) [inline integrations/pagerduty/CHANGELOG.md]:
```suggestion
## 0.5.0 (2026-02-12)
```
